### DISCUSSION TITLE
How to attach multi hooks to an event?

### DISCUSSION THREAD

**Discussion by cigarl** [cilium/ebpf#1468] [Q&A]
I want to attach three or more hooks in one event,but i don't know how to load prog in user space by golang.

1. Do I need to call `loadBpfObjects` and `link.xxx` once for each hook in `main.go`? So what's the difference between that and splitting it into multiple events?
2. How to share the `MAP` in multi hooks? Is the same `MAP` used by multi hooks?
3. Can i attach multi hooks of different types? Such as attach to `kprobe` and `tracepoint` in one event.

I found some examples of kernel space, but no examples of user space.
Any help would be appreciated!

By the way, how do I know what the parameters are for each type of hook function? Such as uretprobe

**Comment by ti-mo** [comment]
I can only answer your question partly, since you're omitting some key info and I'm not sure what you mean by 'event':

1. You only need to call `loadBpfObjects` once, you can attach one instance of a bpf program to as many hooks as you'd like (e.g. attach one kprobe-type program to n amount of kernel symbols)
2. As a rule of thumb, within a Collection, all Maps are shared. If you call `loadBpfObjects` again, that will yield a new Collection, which generally means it gets a new set of Maps. The exception is maps pinned to bpffs, that allows multiple ELF loads to share a single instance of a Map.
3. That again depends, attaching a kprobe-type program to a tracepoint won't work as they require different program types.
